src/rust/container/lib.rs

```diff
@@ -0,0 +1,347 @@
+    messages_callback: Pin<&'static mut ffi::MessageCallback>,
+) -> Result<Box<ContainerService>, ContainerError> {
+    let r#impl = Arc::new(Impl::new(address, container_name, messages_callback)?);
+    let service = Box::new(ContainerService { r#impl });
+    Ok(service)
+}
+
+impl ContainerService {
+    pub fn write_data(&mut self, data: &[u8]) -> bool {
+        if self
+            .r#impl
+            .write_shutdown
+            .load(std::sync::atomic::Ordering::Relaxed)
+        {
+            return false;
+        }
+        self.r#impl.sender.try_send(data.to_vec()).is_ok()
+    }
+
+    pub fn shutdown_write(&mut self) {
+        self.r#impl
+            .write_shutdown
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+
+    #[must_use]
+    pub fn is_write_disconnected(&self) -> bool {
+        self.r#impl
+            .write_shutdown
+            .load(std::sync::atomic::Ordering::Relaxed)
+    }
+}
+
+struct Impl {
+    sender: mpsc::SyncSender<Vec<u8>>,
+    write_shutdown: std::sync::atomic::AtomicBool,
+}
+
+impl Impl {
+    pub fn new(
+        address: &str,
+        container_name: &str,
+        mut messages_callback: Pin<&'static mut ffi::MessageCallback>,
+    ) -> Result<Self, ContainerError> {
+        let (input_sender, input_receiver) = mpsc::sync_channel(1000);
+        let (output_sender, output_receiver) = mpsc::sync_channel::<Vec<u8>>(1000);
+
+        let server = Server::connect(address, container_name)?;
+
+        cxx_integration::tokio::spawn(async move {
+            #[allow(clippy::let_underscore_future)]
+            let _ = tokio::task::LocalSet::new().run_until(async move {
+                let client: container::Client = capnp_rpc::new_client(server);
+                let network = capnp_rpc::twoparty::VatNetwork::new(
+                    MpscReader::new(input_receiver),
+                    MpscWriter::new(output_sender),
+                    rpc_twoparty_capnp::Side::Server,
+                    ReaderOptions::default(),
+                );
+                let rpc_system = capnp_rpc::RpcSystem::new(Box::new(network), Some(client.client));
+
+                tokio::task::spawn_local(rpc_system);
+            });
+
+            while let Ok(message) = output_receiver.recv() {
+                messages_callback.as_mut().call(&message);
+            }
+        });
+
+        Ok(Impl {
+            sender: input_sender,
+            write_shutdown: std::sync::atomic::AtomicBool::new(false),
+        })
+    }
+}
+
+struct Server {
+    docker: Arc<Docker>,
+    container_name: String,
+}
+
+impl Server {
+    fn connect(address: &str, container_name: &str) -> Result<Self, ContainerError> {
+        let docker = Docker::connect_with_socket(address, 120, API_DEFAULT_VERSION)?;
+        Ok(Server {
+            docker: docker.into(),
+            container_name: container_name.to_owned(),
+        })
+    }
+}
+
+impl container::Server for Server {
+    fn destroy(
```

Comment:
I believe what happens here is that we send a SIGTERM, wait for 5 min for the container to stop, and if it is not stopped then we send a SIGKILL.

So I think if this were changed to sending a SIGTERM first, then we waited and checked for status and then called `remove_container` conditionally it'd be the right logic.

Reply:
Oh... maybe stop_container has the behavior we want? (https://docs.rs/bollard/latest/bollard/struct.Docker.html#method.stop_container) and it includes a timeout to wait before it sigkills (https://docs.rs/bollard/latest/bollard/container/struct.StopContainerOptions.html).

So I think:
- If `destroy` is meant to immediately destory, then we call kill_container
- If `destory` does in fact do the timeout behavior, we use stop_container and set the timeout

Reply:
When `signal()` is call, we just relay that signal to docker. When destroy() is called, we should send a SIGKILL.